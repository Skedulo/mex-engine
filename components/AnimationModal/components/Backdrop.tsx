
import React, {forwardRef, useEffect, useImperativeHandle, useRef} from 'react';
import {StyleSheet, TouchableOpacity, Animated} from 'react-native';
import ThemeManager from "../../../mex/colors/ThemeManager";

type BackdropProps = {
    backgroundColor?: string,
    opacity?: number,
    animationDuration?: number,
    visible?: boolean,
    useNativeDriver?: boolean,
    onPress?:  () => void,
    pointerEvents?: "box-none" | "none" | "box-only" | "auto"
};

const Backdrop = (props : BackdropProps, ref:any) => {
    let colors = ThemeManager.getColorSet()
    const animated = useRef(new Animated.Value(0)).current;
    const renderCount = useRef(0);
    const { onPress , pointerEvents = "none", backgroundColor= colors.overlayColor, visible = false, useNativeDriver = false, opacity = 0.5, animationDuration } = props;

    const setOpacity = (initialValue: number) => {
        animated.setValue(initialValue)
    }

    useImperativeHandle(ref, () => ({setOpacity}));

    useEffect(() => {
        renderCount.current = 1;
        if(renderCount.current) return;
        const toValue = visible ? opacity : 0;
        Animated.timing(animated, {
            toValue,
            duration: animationDuration,
            useNativeDriver,
            delay: visible ? 200 : 0
        }).start();
    }, [visible])

    return (
        <Animated.View
            pointerEvents={pointerEvents}
            style={[StyleSheet.absoluteFill, {
                backgroundColor,
                opacity,
            }]}>
            <TouchableOpacity
                onPress={onPress}
                style={StyleSheet.absoluteFill}
            />
        </Animated.View>
    )
};

export default forwardRef(Backdrop);