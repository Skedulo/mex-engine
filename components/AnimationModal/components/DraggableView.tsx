import React, {useRef} from 'react';
import {Animated, LayoutChangeEvent, StyleProp} from 'react-native';

type DraggableViewProps = {
    style?: StyleProp<any>,
    children?:any
};

const DraggableView = (props: DraggableViewProps) => {
    const { style, children: renderContent } = props;
    const layout = useRef<any>({});

    const onLayout = (event: LayoutChangeEvent) => {
        layout.current = event.nativeEvent.layout;
    }

    const content = renderContent({
        onLayout: onLayout,
    });

    return (
        <Animated.View style={style}>
            {content}
        </Animated.View>
    );
}

export default DraggableView;
