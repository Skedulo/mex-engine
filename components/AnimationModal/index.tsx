import Modal from './views/Modal';
import BottomModal from './views/BottomModal';

export {
    Modal,
    BottomModal
};
