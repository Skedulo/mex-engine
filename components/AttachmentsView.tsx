import {UseAttachments} from "@skedulo/mex-types";
import {useEffect, useReducer, useRef} from "react";
import {AttachmentMetadata} from "../../mex-engine-proxy/src";
import lodash from "lodash";
import AttachmentsManager from "../mex/common/attachments/AttachmentsManager";
import {View} from "react-native";
import SkeduloImage from "./SkeduloImage";
import * as React from "react";

type AttachmentsViewProps = {
    dataContext: any,
    jsonDef: UseAttachments,
    styles?: AttachmentsStyle
}

type AttachmentsStyle = {
    borderRadius?: number,
    aspectRatio?: number,
    width?: number,
    marginRight?: number,
    marginTop?: number,
}

export const AttachmentsView = (props: AttachmentsViewProps) => {
    let dataContext = props.dataContext
    let { jsonDef, styles } = props

    const [, forceUpdate] = useReducer(x => x + 1, 0);

    let attachmentsMetadata = useRef<AttachmentMetadata[]>([])
    let attachmentsRef = useRef()

    // subscribe for changes
    useEffect(() => {

        let getAttachmentsDebounce = lodash.throttle(function(attachments: AttachmentMetadata[]) {
            attachmentsMetadata.current = attachments;

            forceUpdate();
        }, 500, {'leading': false})

        AttachmentsManager.observeAttachmentsChangeForContext(attachmentsRef, dataContext.item.UID, "ATTACHMENT", jsonDef.categoryName ?? "", getAttachmentsDebounce);

        return () => {
            AttachmentsManager.unsubscribeAttachmentsChangeForContext(attachmentsRef, dataContext.item.UID, "ATTACHMENT", jsonDef.categoryName ?? "")
        }
    }, [])

    if (attachmentsMetadata.current.length == 0) {
        return <></>
    }

    let item = attachmentsMetadata.current
        .sort((a: AttachmentMetadata, b:AttachmentMetadata) => a.uploadDate < b.uploadDate ? 1 : -1)[0]

    let finalUrl = item.localFileURL ?? item.downloadURL;

    return (
        <View style={[{
            borderRadius: 10,
            aspectRatio: 16 / 9,
        }, styles ?? {}]}>
            <SkeduloImage
                style={{
                    height: "100%",
                    width: "100%",
                }}
                imageStyles={{
                    borderRadius: styles?.borderRadius ?? undefined
                }}
                uri={finalUrl}/>
        </View>)
}
