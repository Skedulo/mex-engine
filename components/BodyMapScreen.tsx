// import React, {useEffect, useState} from 'react'
// import {ActivityIndicator, SafeAreaView} from "react-native";
// import LinearGradient from 'react-native-linear-gradient';
// import {DrawWithOptions} from "./Bodymap";
// import NavigationProcessManager from "../mex/common/NavigationProcessManager";
// import AssetsManager from "../mex/assets/AssetsManager";
// import ThemeManager from "../mex/colors/ThemeManager";
//
// type Props = {
//     route: any,
//     navigation: any
// }
//
// const BodyMapScreen: React.FC<Props> = ({ route}) => {
//
//     let { imageSource } = route.params.pageData;
//     let [accessToken, setAccessToken] = useState('')
//
//     if (typeof imageSource === 'string' || imageSource instanceof String) {
//         imageSource = {
//             uri: imageSource,
//             headers: {
//                 Authorization: 'Bearer ' + accessToken
//             }
//         }
//
//         useEffect(() => {
//             AssetsManager.getAccessToken()
//                 .then((accessToken) => {
//                     setAccessToken(accessToken)
//                 })
//         }, [imageSource])
//     } else {
//         accessToken = "dummy"
//     }
//
//     const closePage = function() {
//         NavigationProcessManager.goBack( null)
//     }
//
//     const takeSnapShotAndLeave = function(uri: string|undefined) {
//         NavigationProcessManager.goBack(uri)
//     }
//
//     return (
//         <SafeAreaView style={{flex: 1}}>
//             {accessToken
//             ?   <DrawWithOptions
//                     linearGradient={LinearGradient}
//                     image={imageSource}
//                     close={closePage}
//                     takeSnapshot={snap => {
//                         snap.then(takeSnapShotAndLeave);
//                     }}
//                 />
//                 : <ActivityIndicator
//                     style={{position: "absolute"}}
//                     color={ThemeManager.getColorSet().skedBlue900} size="small" />}
//
//         </SafeAreaView>
//     )
// }
//
// export default BodyMapScreen
