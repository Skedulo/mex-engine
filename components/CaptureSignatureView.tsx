import React from 'react';
import {requireNativeComponent, ViewProps} from 'react-native';

const RCTSignatureView = requireNativeComponent<Props>('RCTSignatureView');

type Props = ViewProps & {
    onSignatureSaved: (data:OnSignatureSavedResult) => void,
    onSignatureChanged: (data:OnSignatureChangedResult) => void,
}

type OnSignatureSavedResult = {
    imageUrl: string
}

type OnSignatureChangedResult = {
    hasDrawing: boolean
}

const CaptureSignatureView = React.forwardRef((props: Props, ref) => {
    return  <RCTSignatureView {...props} ref={ref as any}/>
})

export default CaptureSignatureView
