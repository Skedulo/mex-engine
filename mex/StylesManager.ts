import ThemeManager from './colors/ThemeManager'
import {Platform} from "react-native";

type StyleConstant = {
    defaultHorizontalPadding: number
    defaultVerticalPadding: number,
    smallVerticalPadding: number,
    componentVerticalPadding: number,
    betweenTextSpacing: number,
    captionTextSize: number,
    betweenComponentVerticalSpacing: number
}

class StylesManager {
    styles: any;
    stylesConstant: StyleConstant = {
        defaultHorizontalPadding: 16,
        defaultVerticalPadding: 16,
        smallVerticalPadding: 8,
        componentVerticalPadding: 12,
        betweenTextSpacing: 6,
        captionTextSize: 12,
        betweenComponentVerticalSpacing: 32
    };

    initializeStyles() {
        let colors = ThemeManager.getColorSet()

       this.styles =  {
           text: {
               "color": colors.skeduloText,
               "fontSize": 20,
           },

           errorText: {
               "color": colors.red800,
               "fontSize": 14,
           },
           textRegular: {
               fontWeight: "normal",
               fontSize: 16,
               flexWrap: 'wrap',
               color: colors.skeduloText
           },
           textTitleListItem: {
               fontWeight: "600",
               fontSize: 16,
               color: colors.skeduloText
           },
           textCaptionListItem: {
               fontWeight: "normal",
               fontSize: 16,
               color: colors.navy600
           },
           textCaptionSelector: {
               fontWeight: "normal",
               fontSize: 14,
               color: colors.navy600
           },
           textCaption: {
               fontWeight: "normal",
               fontSize: 12,
               flexWrap: 'wrap',
               color: colors.skeduloText
           },
           textMedium: {
               fontWeight: "500",
               fontSize: 16,
               color: colors.skeduloText
           },

           textHeadingBold: {
               fontWeight: "bold",
               fontSize: 18,
               color: colors.skeduloText
           },

           navigationTitle: {
               fontSize: 16,
               fontWeight: "bold"
           },

           navigationSubtitle: {
               fontSize: 14
           },
           editText: {
               borderColor: colors.navy300,
               borderWidth: 1,
               borderRadius: 2,
               marginTop: 8,
               height: 48,
               fontSize: 16,
               paddingHorizontal: 16,
               justifyContent: 'center',
               color: colors.skeduloText,
           },
           selector: {
               borderWidth: 1,
               borderRadius: 2,
               marginTop: 8,
               height: 48,
               paddingHorizontal: 16,
           },
           searchBarTextContainer: {
               borderColor: colors.navy100,
               backgroundColor: colors.navy75,
               borderWidth: 1,
               borderRadius: 10,
               paddingHorizontal: 12,
           },
           searchBarText: {
               backgroundColor: "transparent",
               borderWidth: 0,
               borderRadius: 0,
               paddingHorizontal: 12,
               paddingTop: 12,
               paddingBottom: 12,
               color: colors.skeduloText
           },
           button: {
               backgroundColor: colors.skeduloBlue,
               height: 44,
               borderRadius: 8,
               paddingVertical: 10,
               paddingHorizontal: 12,
               justifyContent: "center"
           },
           buttonSmall: {
               backgroundColor: colors.skeduloBlue,
               borderRadius: 8,
               justifyContent: "center",
               height: 32,
               paddingVertical: 8,
               paddingHorizontal: 20
           },
           buttonDefault: {
               height: 44,
               borderRadius: 8,
               paddingVertical: 10,
               paddingHorizontal: 12,
               justifyContent: "center",
               borderColor: colors.skeduloBlue,
               borderWidth: 1
           },
           headerButtonText: {
               fontSize: 16,
               color: colors.white,
               fontWeight: 'normal',
               alignSelf: 'center',
           },
           buttonInsideText: {
               fontSize: 16,
               color: colors.white,
               fontWeight: "600",
               alignSelf: "center",
           },
           buttonSmallInsideText: {
               fontSize: 14,
               color: colors.white,
               fontWeight: "600",
               alignSelf: "center",
           },
           buttonDefaultInsideText: {
               fontSize: 16,
               color: colors.skeduloBlue,
               fontWeight: "500",
               alignSelf: "center",
           },

           buttonLinkText: {
               fontSize: 16,
               color: colors.skeduloText,
               fontWeight: "500",
               alignSelf: "center",
           },

           headerButton: {
               color: colors.white
           },
           headerButtonAlt: {
               color: colors.skedBlue800
           },
           checkBoxProps: {
               size: 25,
               useNativeDriver: true,
               unfillColor: colors.white,
               disableBuiltInState: true,
               iconStyle: {
                   borderRadius: Platform.OS === 'ios' ? 13 : 5
               }
           },
           headerBarSelectView: {
               borderRadius: 8,
               flexDirection: 'row',
               justifyContent:'space-between',
               paddingTop: 19,
               paddingBottom:15,
               paddingHorizontal: 16
           }
       }
    }

    getStyles() { return this.styles }
    getStyleConst() : StyleConstant { return this.stylesConstant }

}

export default new StylesManager() as StylesManager

