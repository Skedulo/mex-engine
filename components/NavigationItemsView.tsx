import * as React from "react";
import StylesManager from "../mex/StylesManager"
import {TouchableOpacity, View, Image, ImageSourcePropType, Text} from "react-native";

export type NavigationItemsViewProps = {
    items: RightNavigationItem[]
}

export type RightNavigationItem = {
    onClicked: (() => void) | undefined
    iconSource?: (() => ImageSourcePropType)
    text?: string
}

export let NavigationItemsView: React.FC<NavigationItemsViewProps> = (props) => {
    let styles = StylesManager.getStyles()
    const getHitSlop = (numberOfItems: number, itemIndex: number) => {
        if (numberOfItems === 1) { // in case there is only 1 button
            return {left: 20, right: 20, top: 20, bottom: 20}
        }
        if (itemIndex === 0) { // there is more than 1 button, and this is the first button (ltr)
            return {left: 20, right: 4, top: 20, bottom: 20}
        }
        if (itemIndex === numberOfItems - 1) { // the last button (ltr)
            return {left: 4, right: 20, top: 20, bottom: 20}
        }
        return {left: 4, right: 4, top: 20, bottom: 20}
    }

    return (
        <View style={{
            flexDirection: 'row',
            flexWrap: 'wrap'
        }}>
            {props.items.map((item, index) => {
                return (
                    <TouchableOpacity
                        onPress={() => item.onClicked?.()}
                        key={index}
                        hitSlop={getHitSlop(props.items.length, index)}
                        style={{ marginLeft: 8, justifyContent: "center" }}>
                        {item.iconSource === undefined
                            ? (<Text style={{...styles.headerButtonText}}>{item.text}</Text>)
                            : (<Image
                                source={item.iconSource()}
                                style={{height: 25, width: 25}}/>)}
                    </TouchableOpacity>
                )
            })}
        </View>)
}
