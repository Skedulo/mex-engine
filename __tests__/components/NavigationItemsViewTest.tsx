import React from "react";
import StylesManager from "../../mex/StylesManager";

import {
    render,
} from "@testing-library/react-native";
import {NavigationItemsView} from "../../components/NavigationItemsView";

describe('NavigationItemsView component tests', function () {

    it('NavigationItemsView basic render', async function () {

        StylesManager.initializeStyles()

        let renderer = render(<NavigationItemsView items={[
            {text: "Test", onClicked: () => {}},
            {iconSource: () => require("../../img/Search_Placeholder.png"), onClicked: () => {}
            }]} />)

        expect(renderer.toJSON()).toMatchSnapshot()
    });

    it('NavigationItemsView empty render', async function () {

        StylesManager.initializeStyles()

        let renderer = render(<NavigationItemsView items={[]} />)

        expect(renderer.toJSON()).toMatchSnapshot()
    });

});
