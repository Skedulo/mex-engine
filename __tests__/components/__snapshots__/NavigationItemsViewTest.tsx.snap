// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationItemsView component tests NavigationItemsView basic render 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "flexWrap": "wrap",
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 20,
        "left": 20,
        "right": 4,
        "top": 20,
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "justifyContent": "center",
        "marginLeft": 8,
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#ffffff",
          "fontSize": 16,
          "fontWeight": "normal",
        }
      }
    >
      Test
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 20,
        "left": 4,
        "right": 20,
        "top": 20,
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "justifyContent": "center",
        "marginLeft": 8,
        "opacity": 1,
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../img/Search_Placeholder.png",
        }
      }
      style={
        Object {
          "height": 25,
          "width": 25,
        }
      }
    />
  </View>
</View>
`;

exports[`NavigationItemsView component tests NavigationItemsView empty render 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "flexWrap": "wrap",
    }
  }
/>
`;
